Sound asset implements

marauder2k7 opened the pull request.

marauder2k7 committed:
SoundAssetImplements

-Explosion
-Lightning
-Splash

-Other sound asset implementations will require soundasset array if possible.

marauder2k7 committed:
Update SoundAsset.h

-SoundAsset array setters getters and binds in place

marauder2k7 committed:
SoundAsset Array

Few fixes to soundasset array to stop it complaining

marauder2k7 committed:
More Implements

-Most Vehicles and FX classes
-Vehicle classes may need more preloads for assets.

dottools commented on Engine/source/T3D/fx/lightning.h:
```diff
@@ -86,6 +92,22 @@ class LightningData : public GameBaseData
+         return NULL;
+   }
+
+   SFXProfile* getSFXProfile() {
```
Uh oh. `One true brace` coding style leaked here...

Also it is better coding style for functions that return values to always have a non-conditional return statement at the end of the function, instead of nested under an else, like so:

```cpp
SFXProfile* getSFXProfile()
{
   if (mStrikeSoundAsset.notNull())
      return mStrikeSoundAsset->getSfxProfile();

   return NULL;
}
```

That way the intent is more clear. If mStrikeSoundAsset is not NULL return it. At end of function it just returns NULL as in failure to satisfy request, operation failed, etc..

marauder2k7 replied:
copied from the way it was done before

marauder2k7 replied:
https://github.com/TorqueGameEngines/Torque3D/blob/Preview4_0/Engine/source/T3D/fx/precipitation.h#L76

dottools commented on Engine/source/T3D/fx/splash.h:
```diff
@@ -91,8 +92,19 @@ class SplashData : public GameBaseData
+   DECLARE_SOUNDASSET_SETGET(SplashData, Sound);
+
+   /// this should probably be added as a function higher up to stop repeats.
+   SFXProfile* getSFXProfile() {
```
Eww it happened again...

marauder2k7 replied:
copied from the way it was done before

marauder2k7 replied:
https://github.com/TorqueGameEngines/Torque3D/blob/Preview4_0/Engine/source/T3D/fx/precipitation.h#L76

dottools commented on Engine/source/T3D/projectile.cpp:
```diff
@@ -368,9 +367,8 @@ bool ProjectileData::preload(bool server, String &errorStr)
-      String sfxErrorStr;
-      if( !sfxResolve( &sound, sfxErrorStr ) )
-         Con::errorf(ConsoleLogEntry::General, "ProjectileData::preload: Invalid packet: %s", sfxErrorStr.c_str());
+      if( !getProjectileSound() )
```
Feel free to delete the erroneous/unnecessary spaces inside parenthesis (right of `(` and left of `)`). Optionally space between keyword and expression (except for operators, such as like `sizeof()`, casts)
example:
```cpp
if( !getProjectileSound() )
```
to
```cpp
if (!getProjectileSound())
```

marauder2k7 replied:
copied from the way it was done before

dottools commented on Engine/source/T3D/projectile.h:
```diff
@@ -115,7 +116,14 @@ class ProjectileData : public GameBaseData
-   SFXTrack* sound;                    // Projectile Sound
+   DECLARE_SOUNDASSET(ProjectileData, ProjectileSound);
+   DECLARE_SOUNDASSET_SETGET(ProjectileData, ProjectileSound);
+   SFXProfile* getSFXProfile() {
```
And again :(

marauder2k7 replied:
copied from the way it was done before

marauder2k7 replied:
https://github.com/TorqueGameEngines/Torque3D/blob/Preview4_0/Engine/source/T3D/fx/precipitation.h#L76

dottools commented on Engine/source/T3D/vehicles/wheeledVehicle.cpp:
```diff
@@ -301,7 +301,7 @@ WheeledVehicleData::WheeledVehicleData()
    dMemset(&wheel, 0, sizeof(wheel));
    for (S32 i = 0; i < MaxSounds; i++)
-      sound[i] = 0;
+      INIT_SOUNDASSET_ARRAY(WheeledVehicleSounds, i);
```
Danger! Do not trust preprocessor to handle code body braces for you in `if ()`, `while ()`, `for ()`, etc.. statements as it is known to be [broken for various compilers and can lead to bugs](https://stackoverflow.com/questions/154136/why-use-apparently-meaningless-do-while-and-if-else-statements-in-macros). Always explicitly put macros inside of code body braces:
```cpp
for (S32 i = 0; i < MaxSounds; i++)
{
   INIT_SOUNDASSET_ARRAY(WheeledVehicleSounds, i);
}
```

Even though `INIT_SOUNDASSET_ARRAY` macro definition is code body braced doesn't mean all preprocessors will handle it well. And for others the classic annoying unnecessary trailing ; warning message.

marauder2k7 replied:
copied from the way it was done before

marauder2k7 committed:
More turns

-Flying Vehicle
-Proximinity Mine

marauder2k7 committed:
Preload all soundAsset

-Fixed: Preloads need to preload all soundAssets not just the ones it was preloading before.

marauder2k7 committed:
QOC

Code cleaning of unnecessary else'
Remove unnecessary dynamic_casts

Azaezel commented:
git mangled this PR, so we followed up via #616
